### ISSUE TITLE
No avcodec in java.library.path

### ISSUE THREAD

**Issue by felipeek** [bytedeco/javacv#80]
Hi, I'm developing a Java Web Application. I'm using JavaCV to capture single frames from videos, which will be uploaded by the user. This frames will serve as placeholders to the user.

I'm developing the application in the local runtime and everything is working 100% perfectly. The JavaCV is working correctly and doing everything I want. However, my problem happens when I try to publish my application in an external server.

I keep getting this error:

`java.lang.UnsatisfiedLinkError: no avcodec in java.library.path`

Again, this error only occurs in the real server. In my local runtime, everything is fine.

Additional Information:

I'm using JavaCV 0.8, and this is my maven declaration:

``` xml
<!-- JavaCV -->
<dependency>
    <groupId>org.bytedeco</groupId>
    <artifactId>javacv</artifactId>
    <version>0.8</version>
</dependency>
<dependency>
    <groupId>org.bytedeco</groupId>
    <artifactId>javacpp</artifactId>
    <version>0.8</version>
</dependency>
```

Firstly, I wasn't using the javacpp dependency. I added it later, but it isn't working both ways.

This is the code that I'm using to work with JavaCV:

``` java
FFmpegFrameGrabber fg = new FFmpegFrameGrabber(f);
fg.start();

ByteArrayOutputStream baos = new ByteArrayOutputStream();

ImageIO.write(fg.grab().getBufferedImage(), "png", baos);

fg.stop();
```

This is the error I'm getting:

``` text
Servlet.service() for servlet [mvc-dispatcher] in context with path [/ArrivalGuideCMF] threw exception [Handler processing failed; nested exception is java.lang.UnsatisfiedLinkError: no jniavcodec in java.library.path] with root cause

java.lang.UnsatisfiedLinkError: no avcodec in java.library.path

    at java.lang.ClassLoader.loadLibrary(ClassLoader.java:2018)
    at java.lang.Runtime.loadLibrary0(Runtime.java:845)
    at java.lang.System.loadLibrary(System.java:1091)
    at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:524)
    at org.bytedeco.javacpp.Loader.load(Loader.java:390)
    at org.bytedeco.javacpp.Loader.load(Loader.java:353)
    at org.bytedeco.javacpp.avcodec.<clinit>(avcodec.java:12)
    at org.bytedeco.javacv.FFmpegFrameGrabber.<clinit>(FFmpegFrameGrabber.java:103)
        ...
```

I only have the maven dependency to run JavaCV, no other external JARs are linked to my project. However, I've already tried to add ffmpeg-20130915-git-7ac6c63-windows-x86_64.jar and ffmpeg-20130915-git-7ac6c63-windows-x86.jar to my project, as I saw in this forum, but it didn't work.

One more thing: I'm using JavaCV 0.8, since this is the only version that was working for me (at least in my local runtime). JavaCV 10.0 and older versions are not working and are giving me more errors.

Any help is appreciated. I ensure you I've searched A LOT how to fix it, but nothing is working!

The server I'm publishing my application is the HCP (Hana Cloud Platform) Trial, which accepts Java Applications (just to you know)

**Comment by saudet** [comment]
It's always interesting to hear about new applications, thanks! :)

It simply looks like the native libraries are not getting loaded. How did you define the `platform.dependencies` system property for Maven?

**Comment by felipeek** [comment]
@saudet Thanks for the response. I'm running the following maven command on my project:

```text
mvn package -Dplatform.dependencies=true
```

In my pom.xml, I only have the JavaCV dependency:

```xml
<!-- JavaCV -->
    <dependency>
    <groupId>org.bytedeco</groupId>
    <artifactId>javacv</artifactId>
    <version>0.8</version>
</dependency>
```

I'm not very conversant with Maven, so maybe I forgot something. If you could help me with that, I would be grateful.

Thanks

**Comment by saudet** [comment]
That's alright, that should pull in the native libraries.

I've just remembered that for these kinds of Java containers we need to follow special instructions when using JNI. Do you have anything about JNI in HCP's documentation?

**Comment by felipeek** [comment]
@saudet I've searched in the documentation and I found nothing about JNI, unfortunately. However, the local runtime that I'm using is the HCP local runtime, which makes me believe that if my application is working locally, it should be working in the real HCP server too (unless some files are missing there).

I'm kinda stuck here, do you have any ideas? Do you think removing the maven dependency and adding all the dependencies manually could solve the problem?

Thanks

**Comment by saudet** [comment]
It's not about local or not local, it's about class loaders. These kinds of containers usually make use of more than one class loader, and JNI has issues with that. So, for example, TomCat recommends to use the "shared class loader" for libraries using JNI:
http://wiki.apache.org/tomcat/HowTo#I.27m_encountering_classloader_problems_when_using_JNI_under_Tomcat
There should be something similar in HCP.

It's also possible that the `platform.dependencies` system property doesn't get set on the server, and specifying all the platform-specific dependencies manually could help yes.

**Comment by saudet** [comment]
So, have you figured out what they issue was?

**Comment by saudet** [comment]
I've just released version 1.1. Could you try it out and let me know if this problem still happens? Thanks!

**Comment by saudet** [comment]
The cause of this issue is probably missing dependencies:
https://stackoverflow.com/questions/37080600/web-application-using-javacv-and-deployed-on-sap-hana-cloud-platform-error
This should no longer occur with recent versions of JavaCV and the `javacv-platform` artifact, so I'll mark this as resolved, but please let me know if you still have issues. Thanks!
